consensus.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+// GetVoteChoices returns configured vote preferences for each agenda of the
+// latest supported stake version for the wallet with the specified walletID.
+func (mw *MultiWallet) GetVoteChoices(hash string, walletID int) (*GetVoteChoicesResult, error) {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return nil, errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]AgendaVoteChoice, len(agendas)),
+	}
+
+	ctx := wallet.shutdownContext()
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		agenda := mw.agenda(agendas, choices[i].AgendaID)
+		resp.Choices[i] = AgendaVoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agenda.Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for _, choice := range agenda.Vote.Choices {
+			if choices[i].ChoiceID == choice.Id {
+				resp.Choices[i].ChoiceDescription = choice.Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+func (mw *MultiWallet) agenda(agendas []chaincfg.ConsensusDeployment, agendaID string) *chaincfg.ConsensusDeployment {
+	for _, agenda := range agendas {
+		if agenda.Vote.Id == agendaID {
+			return &agenda
+		}
+	}
+
+	return nil
+}
+
+// SetVoteChoice sets a voting choice for the specified agenda using the
+// specified ticket hash controlled by a VSP.
+//
+// If a ticket hash isn't provided, the vote choice is set for all tickets
+// controlled by the set VSP.
+func (mw *MultiWallet) SetVoteChoice(walletID int, vspHost string, vspPubKey []byte, agendaID, choiceID, hash string, passphrase []byte) error {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	if vspHost == "" {
+		return fmt.Errorf("request requires a vspHost but no vspHost was provided")
+	}
+
+	// Test to ensure the passphrase is correct, and the voting is authorized.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
+	if err != nil {
+		return err
+	}
+
+	vspClient, err := wallet.VSPClient(vspHost, vspPubKey)
+	if err != nil {
+		return err
+	}
```

Comment:
vspClient is recreated in the if block below, this will never be used if ticketHash is not nil. It is only used if ticketHash is nil, after the if block below. This assignment should move to after the `if ticketHash != nil {` block.